File: ported/auto/9bbb9b.js
// Auto-translated from Ghidra C by tools/c-to-js/translate.js.
// Source: decompiled/c/9bbb9b.c
// Edit by hand only after diff-test passes — re-running the translator will overwrite.

/** @typedef {import("../../runtime/heap.js").Heap} Heap */

import { CONCAT11, CONCAT22, CONCAT31 } from "../../runtime/ghidra-builtins.js";
import { regs } from "../../runtime/regs.js";
import { FUN_00408342 } from "./408342.js";
import { FUN_00408387 } from "./408387.js";
import { FUN_004083b5 } from "./4083b5.js";
import { FUN_004083e1 } from "./4083e1.js";
import { FUN_00458bcf } from "./458bcf.js";
export function FUN_009bbb9b(heap) {
  let cVar1 = 0;
  let sVar2 = 0;
  let uVar3 = 0;
  let iVar4 = 0;
  let iVar5 = 0;
  let uVar6 = 0;
  let sVar7 = 0;
  let cVar8 = 0;
  let uVar9 = 0;
  let pcVar10 = 0;
  let puVar11 = 0;
  let pcVar12 = 0;
  let puVar13 = 0;
  LAB_009bbdbf: {
  LAB_009bbdbe: {
  sVar2 = ((1) & 0xffff);
  while (true) {
    pcVar12 = ((0x0099a888) >>> 0);
    pcVar10 = ((0x005f8da3) >>> 0);
    heap.setU16((0x00971e86 + 0), (sVar2) & 0xffff);
    do {
      cVar1 = ((heap.i8(pcVar10)) & 0xff);
      heap.setU32(pcVar12, (cVar1) & 0xffffffff);
      pcVar10 = ((pcVar10 + 1) >>> 0);
      pcVar12 = ((pcVar12 + 1) >>> 0);
    } while (cVar1 != 0);
    (regs.eax = FUN_00458bcf(heap));
    iVar4 = (((regs.eax = FUN_004083b5(heap, 0x0099a888))) >>> 0);
    if ((iVar4 | 0) == -1) {
      break;
    }
    (regs.eax = FUN_00408387(heap, iVar4));
    sVar2 = ((sVar2 + 1) & 0xffff);
  }
  iVar4 = (((regs.eax = FUN_004083e1(heap, 0x0099a888))) >>> 0);
  if ((iVar4 | 0) != -1) {
    puVar11 = ((0x00981efc) >>> 0);
    sVar7 = ((0x80) & 0xffff);
    heap.setU32(0x009a2004, (iVar4) >>> 0);
    do {
      heap.setU32(puVar11, (0) & 0xffffffff);
      puVar11 = ((puVar11 + 1) >>> 0);
      sVar7 = ((sVar7 + -1) & 0xffff);
    } while (sVar7 != 0);
    heap.setU32(0x00981efc, (0x801050a) >>> 0);
    heap.setU32(0x00981f04, (CONCAT22(heap.u32(0x00971ed8) - 1, heap.u32(0x00971ed6) - 1)) >>> 0);
    heap.setU32(0x00981f08, (0x400040) >>> 0);
    heap.setU32(0x00981f3c, (CONCAT22(heap.u32(0x00971ed6), 0x100)) >>> 0);
    iVar4 = (((regs.eax = FUN_00408342(heap, heap.u32(0x009a2004), 0x00981efc, 0x80))) >>> 0);
    if ((iVar4 | 0) != -1) {
      pcVar12 = ((0x00981efc) >>> 0);
      uVar9 = ((((heap.u32(0x00971ed6)) >>> 0)) >>> 0);
      iVar4 = ((heap.u32(0x00971ed8) * uVar9) >>> 0);
      pcVar10 = ((heap.u32(0x0099fb7c)) >>> 0);
      while (true) {
        if (uVar9 == 0) {
          pcVar10 = ((pcVar10 + (((heap.u32(0x0099fb88) - heap.u32(0x00971ed6)) + heap.u32(0x0099fb84)) & 0xffff)) >>> 0);
          uVar9 = ((((heap.u32(0x00971ed6)) >>> 0)) >>> 0);
        }
        if (iVar4 == 0) {
          break;
        }
        if (0x991e97 < pcVar12) {
          iVar5 = (((regs.eax = FUN_00408342(heap, heap.u32(0x009a2004), 0x00981efc, pcVar12 + -0x981efc))) >>> 0);
          if ((iVar5 | 0) == -1) {
            /* goto LAB_009bbdb0 — unsupported, early-return */ if (typeof globalThis._gotoWarn !== 'undefined') globalThis._gotoWarn("FUN_009bbb9b/LAB_009bbdb0"); return 0;
          }
          pcVar12 = ((0x00981efc) >>> 0);
        }
        cVar1 = ((heap.i8(pcVar10)) & 0xff);
        pcVar10 = ((pcVar10 + 1) >>> 0);
        iVar4 = ((iVar4 + -1) >>> 0);
        uVar9 = ((uVar9 - 1) >>> 0);
        uVar3 = ((CONCAT11(cVar1, cVar1) & 0xc0ff) & 0xffff);
        cVar1 = ((((uVar3) << 24 >> 24)) & 0xff);
        if (((((uVar3 >>> 8)) << 24 >> 24) == -0x40) || ((uVar9 != 0 && (cVar1 == heap.i8(pcVar10))))) {
          cVar8 = ((-0x3f) & 0xff);
          for (; ((uVar9 != 0 && ((cVar8 | 0) != -1)) && (cVar1 == heap.i8(pcVar10))); pcVar10 = (((pcVar10 + 1) >>> 0)) >>> 0) {
            iVar4 = ((iVar4 + -1) >>> 0);
            uVar9 = ((uVar9 - 1) >>> 0);
            cVar8 = ((cVar8 + 1) & 0xff);
          }
          heap.setU32(pcVar12, (cVar8) & 0xffffffff);
          heap.setI8((pcVar12 + (1)), (cVar1) & 0xff);
          pcVar12 = ((pcVar12 + 2) >>> 0);
        } else {
          heap.setU32(pcVar12, (cVar1) & 0xffffffff);
          pcVar12 = ((pcVar12 + 1) >>> 0);
        }
      }
      if (pcVar12 + -0x981efc != 0x0) {
        iVar4 = (((regs.eax = FUN_00408342(heap, heap.u32(0x009a2004), 0x00981efc, pcVar12 + -0x981efc))) >>> 0);
        if ((iVar4 | 0) == -1) {
          /* goto LAB_009bbdb0 — unsupported, early-return */ if (typeof globalThis._gotoWarn !== 'undefined') globalThis._gotoWarn("FUN_009bbb9b/LAB_009bbdb0"); return 0;
        }
      }
      heap.setU32(0x00981efc, (CONCAT31(heap.u32(0x00981efd), 0xc)) >>> 0);
      puVar13 = ((0x00981efd) >>> 0);
      puVar11 = ((0x005f2000) >>> 0);
      sVar7 = ((0x100) & 0xffff);
      do {
        heap.setU8((puVar13 + (2)), (heap.u8(puVar11)) & 0xff);
        heap.setU8((puVar13 + (1)), (heap.u8(puVar11 + (1))) & 0xff);
        heap.setU32(puVar13, (heap.u8(puVar11 + (2))) & 0xffffffff);
        puVar11 = ((puVar11 + 4) >>> 0);
        puVar13 = ((puVar13 + 3) >>> 0);
        sVar7 = ((sVar7 + -1) & 0xffff);
      } while (sVar7 != 0);
      iVar4 = (((regs.eax = FUN_00408342(heap, heap.u32(0x009a2004), 0x00981efc, 0x301))) >>> 0);
      if ((iVar4 | 0) != -1) {
        iVar4 = (((regs.eax = FUN_00408387(heap, heap.u32(0x009a2004)))) >>> 0);
        if ((iVar4 | 0) != -1) {
          uVar6 = ((((((iVar4) >>> 0) >>> 0x10) & 0xffff)) & 0xffff);
          break LAB_009bbdbf;
        }
        break LAB_009bbdbe;
      }
    }
    LAB_009bbdb0: iVar4 = (((regs.eax = FUN_00408387(heap, heap.u32(0x009a2004)))) >>> 0);
  }
  }
  uVar6 = ((((((iVar4) >>> 0) >>> 0x10) & 0xffff)) & 0xffff);
  }
  return CONCAT22(uVar6, sVar2);
}

File: decompiled/c/9bbb9b.c

/* WARNING: Globals starting with '_' overlap smaller symbols at the same address */

undefined4 FUN_009bbb9b(void)

{
  char cVar1;
  short sVar2;
  ushort uVar3;
  int iVar4;
  int iVar5;
  undefined2 uVar6;
  short sVar7;
  char cVar8;
  uint uVar9;
  char *pcVar10;
  undefined1 *puVar11;
  char *pcVar12;
  undefined1 *puVar13;
  
  sVar2 = 1;
  while( true ) {
    pcVar12 = &DAT_0099a888;
    pcVar10 = (char *)&DAT_005f8da3;
    DAT_00971e86._0_2_ = sVar2;
    do {
      cVar1 = *pcVar10;
      *pcVar12 = cVar1;
      pcVar10 = pcVar10 + 1;
      pcVar12 = pcVar12 + 1;
    } while (cVar1 != '\0');
    FUN_00458bcf();
    iVar4 = FUN_004083b5(&DAT_0099a888);
    if (iVar4 == -1) break;
    FUN_00408387(iVar4);
    sVar2 = sVar2 + 1;
  }
  iVar4 = FUN_004083e1(&DAT_0099a888);
  if (iVar4 != -1) {
    puVar11 = &DAT_00981efc;
    sVar7 = 0x80;
    DAT_009a2004 = iVar4;
    do {
      *puVar11 = 0;
      puVar11 = puVar11 + 1;
      sVar7 = sVar7 + -1;
    } while (sVar7 != 0);
    _DAT_00981efc = 0x801050a;
    DAT_00981f04 = CONCAT22(DAT_00971ed8 - 1,DAT_00971ed6 - 1);
    DAT_00981f08 = 0x400040;
    DAT_00981f3c = CONCAT22(DAT_00971ed6,0x100);
    iVar4 = FUN_00408342(DAT_009a2004,&DAT_00981efc,0x80);
    if (iVar4 != -1) {
      pcVar12 = &DAT_00981efc;
      uVar9 = (uint)DAT_00971ed6;
      iVar4 = DAT_00971ed8 * uVar9;
      pcVar10 = DAT_0099fb7c;
      while( true ) {
        if (uVar9 == 0) {
          pcVar10 = pcVar10 + (ushort)((DAT_0099fb88 - DAT_00971ed6) + DAT_0099fb84);
          uVar9 = (uint)DAT_00971ed6;
        }
        if (iVar4 == 0) break;
        if ((char *)0x991e97 < pcVar12) {
          iVar5 = FUN_00408342(DAT_009a2004,&DAT_00981efc,pcVar12 + -0x981efc);
          if (iVar5 == -1) goto LAB_009bbdb0;
          pcVar12 = &DAT_00981efc;
        }
        cVar1 = *pcVar10;
        pcVar10 = pcVar10 + 1;
        iVar4 = iVar4 + -1;
        uVar9 = uVar9 - 1;
        uVar3 = CONCAT11(cVar1,cVar1) & 0xc0ff;
        cVar1 = (char)uVar3;
        if (((char)(uVar3 >> 8) == -0x40) || ((uVar9 != 0 && (cVar1 == *pcVar10)))) {
          cVar8 = -0x3f;
          for (; ((uVar9 != 0 && (cVar8 != -1)) && (cVar1 == *pcVar10)); pcVar10 = pcVar10 + 1) {
            iVar4 = iVar4 + -1;
            uVar9 = uVar9 - 1;
            cVar8 = cVar8 + '\x01';
          }
          *pcVar12 = cVar8;
          pcVar12[1] = cVar1;
          pcVar12 = pcVar12 + 2;
        }
        else {
          *pcVar12 = cVar1;
          pcVar12 = pcVar12 + 1;
        }
      }
      if (pcVar12 + -0x981efc != (char *)0x0) {
        iVar4 = FUN_00408342(DAT_009a2004,&DAT_00981efc,pcVar12 + -0x981efc);
        if (iVar4 == -1) goto LAB_009bbdb0;
      }
      _DAT_00981efc = CONCAT31(_DAT_00981efd,0xc);
      puVar13 = &DAT_00981efd;
      puVar11 = &DAT_005f2000;
      sVar7 = 0x100;
      do {
        puVar13[2] = *puVar11;
        puVar13[1] = puVar11[1];
        *puVar13 = puVar11[2];
        puVar11 = puVar11 + 4;
        puVar13 = puVar13 + 3;
        sVar7 = sVar7 + -1;
      } while (sVar7 != 0);
      iVar4 = FUN_00408342(DAT_009a2004,&DAT_00981efc,0x301);
      if (iVar4 != -1) {
        iVar4 = FUN_00408387(DAT_009a2004);
        if (iVar4 != -1) {
          uVar6 = (undefined2)((uint)iVar4 >> 0x10);
          goto LAB_009bbdbf;
        }
        goto LAB_009bbdbe;
      }
    }
LAB_009bbdb0:
    iVar4 = FUN_00408387(DAT_009a2004);
  }
LAB_009bbdbe:
  uVar6 = (undefined2)((uint)iVar4 >> 0x10);
LAB_009bbdbf:
  return CONCAT22(uVar6,sVar2);
}

